### PR TITLE
New Feature: Constant Interpolated Strings (Initial PR)

### PR DESCRIPTION
**Pull request by kevinsun-dev**
This feature is about enabling string constants to be made out of interpolated strings, where the interpolated string itself is made up of constants.
```
public class C
{
    const string S1 = $"Hello world";
    const string S2 = $"Hello{" "}World";
    const string S3 = $"{S1} Kevin, welcome to the team!";
}
```
Added in this PR:
- Support simple constant interpolated strings
- Updated tests

### CODE REVIEW

**Comment by gafter** [comment]
I would think the simplest way of constructing the resulting string, including handling of escaping, would be to use `string.Format` on the format string and all of the values of the interpolations. I think that could be done after processing everything rather than at the same time, separating the constant folding implementation from the rest of the semantic analysis.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 1]
```diff
@@ -25,6 +25,10 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
             var stringType = GetSpecialType(SpecialType.System_String, diagnostics, node);
             var objectType = GetSpecialType(SpecialType.System_Object, diagnostics, node);
             var intType = GetSpecialType(SpecialType.System_Int32, diagnostics, node);
+
+            // PROTOTYPE follow up with Neal on the implementation
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:29]:
```suggestion
            // PROTOTYPE(CONSTISTR) follow up with Neal on the implementation 
```

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 2]
```diff
@@ -54,6 +58,7 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
                             {
                                 alignment = GenerateConversionForAssignment(intType, BindValue(interpolation.AlignmentClause.Value, diagnostics, Binder.BindValueKind.RValue), diagnostics);
                                 var alignmentConstant = alignment.ConstantValue;
+                                constantEnabled = false;
                                 if (alignmentConstant != null && !alignmentConstant.IsBad)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:62]: One potential benefit of going the `string.Format` route is that it could allow us to consider supporting alignment as that would come for free. We should keep that in mind when discussing that aspect of the implementation.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 3]
```diff
@@ -25,6 +25,10 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
             var stringType = GetSpecialType(SpecialType.System_String, diagnostics, node);
             var objectType = GetSpecialType(SpecialType.System_Object, diagnostics, node);
             var intType = GetSpecialType(SpecialType.System_Int32, diagnostics, node);
+
+            // PROTOTYPE follow up with Neal on the implementation 
+            ConstantValue? resultConstant = null;
+            bool constantEnabled = true;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:31]: Think the code would read better if the names of these paired values matched better. For instance `resultConstant` and `isResultConstant`

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 4]
```diff
@@ -92,20 +98,50 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
                             }
 
                             builder.Add(new BoundStringInsert(interpolation, value, alignment, format, null));
+                            if (value.ConstantValue != null)
+                            {
+                                if (!value.ConstantValue.IsString || value.ConstantValue.IsBad || value.ConstantValue.IsNull)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:103]: Right now the logic to determine if this value is constant is spread out to four places: the format check, the alignment check, the constant check and the type of constant. Consider just combining it into a single check here:

```c#
if (value.ConstantValue?.IsString == true &&
    !value.ConstantValue.IsBad &&
    !value.ConstantValue.IsNull &&
    interpolation.FormatClaues is null && 
    interpolation.AlignmentClause is null) 
{
   ...
}
```

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:103]: Why is `null` special cased here as a non-constant? That is a legal `string` constant value. Please add a PROTOTYPE comment to follow up on that.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:29]: This has been done, you can remove the comment :)

---
In reply to: [447263586](https://github.com/dotnet/roslyn/pull/45396#discussion_r447263586) [](ancestors = 447263586)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:103]: `(ConstantValue)null != ConstantValue.Null`

---
In reply to: [447269087](https://github.com/dotnet/roslyn/pull/45396#discussion_r447269087) [](ancestors = 447269087)

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:103]: The first 3 can be collapsed to a single pattern: `value.ConstantValue is { IsString: true, IsBad: false, IsNull: false }`

---
In reply to: [447268625](https://github.com/dotnet/roslyn/pull/45396#discussion_r447268625) [](ancestors = 447268625)

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:103]: And the second 2 as well: `interpolation is { FormatClauses : null, AlignmentClause: null }`

---
In reply to: [447343672](https://github.com/dotnet/roslyn/pull/45396#discussion_r447343672) [](ancestors = 447343672,447268625)

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs [hunk 5]
```diff
@@ -286,6 +286,14 @@ internal partial class BoundBinaryOperator
         }
     }
 
+    internal partial class BoundInterpolatedString
+    {
+        public override ConstantValue? ConstantValue
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:291]: What actually uses this property?

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs [hunk 6]
```diff
@@ -3990,7 +3990,7 @@ static void F()
         }
 
         [WorkItem(976, "https://github.com/dotnet/roslyn/issues/976")]
-        [Fact]
+        [Fact(Skip = "PROTOTYPE(CONSTISTR)")]
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs:3993]: What is failing about this test? Consider recording that in the skip comment.

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 7]
```diff
@@ -3212,7 +3212,7 @@ void M()
         }
 
         [WorkItem(1065661, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/1065661")]
-        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceVariable)]
+        [Fact(Skip = "PROTOTYPE(CONSTISTR)"), Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceVariable)]
```
Comment (333fred) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:3215]: Same here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 8]
```diff
@@ -3462,6 +3462,33 @@ static void Main()
             source.Append(source1);
             CompileAndVerify(source.ToString(), expectedOutput: "58430604");
         }
+
+        [Fact]
+        public void ConstantInterpolatedStringsSimple()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3467]: Please add a prototype comment for testing invalid scenarios as well.

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:291]: Ah, nevermind.

---
In reply to: [447344161](https://github.com/dotnet/roslyn/pull/45396#discussion_r447344161) [](ancestors = 447344161)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:103]: We can also bail out early if `!constantEnabled`.

---
In reply to: [447343824](https://github.com/dotnet/roslyn/pull/45396#discussion_r447343824) [](ancestors = 447343824,447343672,447268625)

**Comment by 333fred** [comment]
Done review pass (commit 15)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 3 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs]: Thanks for the suggestion, changes made.

---
In reply to: [447267474](https://github.com/dotnet/roslyn/pull/45396#discussion_r447267474) [](ancestors = 447267474)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 2 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs]: Yup, talked it over with Fred. The alignment support for strings that can only be composed of constants may be outweighed by this implementation's default support for ropes.

---
In reply to: [447264298](https://github.com/dotnet/roslyn/pull/45396#discussion_r447264298) [](ancestors = 447264298)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 1 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs]: Indeed, the comment has been removed.

---
In reply to: [447335348](https://github.com/dotnet/roslyn/pull/45396#discussion_r447335348) [](ancestors = 447335348,447263586)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 4 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs]: Thanks for all the suggestions, I've incorporated them into commit 16.

---
In reply to: [447347863](https://github.com/dotnet/roslyn/pull/45396#discussion_r447347863) [](ancestors = 447347863,447343824,447343672,447268625)

Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs]: If we're talking about value.ConstantValue, that was null checked to prevent a whole host of NullPointerExceptions. If we're taking about the extra case in the if statement beneath it, I'd honestly forgotten if it was merely out of an abundance of caution or not. Removed it in commit 16, lets see if it works.

---
In reply to: [447269087](https://github.com/dotnet/roslyn/pull/45396#discussion_r447269087) [](ancestors = 447269087)

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs (hunk 6 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs]: Recorded the reason in the skip comment, is this the correct format?

---
In reply to: [447344626](https://github.com/dotnet/roslyn/pull/45396#discussion_r447344626) [](ancestors = 447344626)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs (hunk 8 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3467]: Added tests for invalid scenarios.

---
In reply to: [447344980](https://github.com/dotnet/roslyn/pull/45396#discussion_r447344980) [](ancestors = 447344980)

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs]: That's fine, thanks.

---
In reply to: [447771212](https://github.com/dotnet/roslyn/pull/45396#discussion_r447771212) [](ancestors = 447771212,447344626)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 9]
```diff
@@ -3462,6 +3462,66 @@ static void Main()
...
+$""{""Level 5""} {""Number 3""}"" --> Level 5 Number 3
+$""{$""{""Spinning Top""}""}"" --> Spinning Top
+$""{""Spinning Top""}"" --> Spinning Top
+$""{S1}"" --> Testing
+$""{F1} the {S2}"" --> Testing the Level 5 Number 3";
+            Assert.Equal(expected, actual);
+        }
+
+        [Fact]
+        public void ConstantInterpolatedStringsError()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3494]: We'll need more tests, but it's a good start. The rest can wait for test plan review.

Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3494]: Think there are some we can add in advance of the test plan review: used in attributes, default parameter values, used across assembly boundaries, used as fields, etc ...  Also the always popular: lang ver checks. 

Those can be done in a separate PR but should get them done before a test plan review
